### DISCUSSION TITLE
Jan: Jan has memory

### DISCUSSION THREAD

**Discussion by lv333ming** [janhq/jan#2212] [Feature Ideas]
Is it possible to add a memory function to JAN?

Currently, conversation history is lost when JAN is closed. It would be beneficial to retain this history when JAN is reopened, making conversations more natural and human-like.

I am also curious if conversing with JAN can improve its knowledge. Can the following factors contribute to the model's knowledge growth?

User feedback during conversations: Does the model learn from its mistakes and improve based on user feedback?
New information and knowledge provided by users: Does the model incorporate this information to expand its knowledge base?
I believe that adding a memory function and incorporating user feedback and information into the model's training process would significantly enhance JAN's capabilities and make it a more valuable tool for users.

**Comment by eckartal** [comment]
Up.

![image](https://github.com/user-attachments/assets/924a1e16-7a75-4cd7-8cf2-0044f9ccc913)

Source: https://x.com/easyldur/status/1833481586249241019

**Reply by rvrstfn** [reply]
Stefano Rivera here. These are random ideas about long term / persistent memory.

Methods for saving the memories:
- every conversation is instantly embedded, in overlapping blocks (fast and stupid method) 
- "function calling" is used for saving relevant memories (ChatGPT does this, apparently)
- every conversation is ingested by night to build a knowledge structure (eg. GraphRAG)
- only the latest conversation is summarized (not good)

Methods for retrieving the memories:
- RAG is performed at every user message, and injected in the system prompt
- the full memory structure is injected in the system prompt at the beginning of the conversation, or randomly sampled
- "function calling" is used for RAG on demand

These are just the ideas of a non-expert!

**Reply by rvrstfn** [reply]
Most exotic:
- nightly LoRA fine-tuning on own conversations
